### PR TITLE
Add advice about keychain credentials

### LINKED ISSUE #50050: brew search yields 401 errors
```
tim@rocketman:Homebrew (master)$ brew search python
app-engine-python                       homebrew/python/matplotlib ✔            homebrew/python/pygame                  homebrew/python/retext                  python ✔                              
boost-python ✔                          homebrew/python/matplotlib-basemap      homebrew/python/pymssql                 homebrew/python/scapy                   python-markdown                       
gst-python                              homebrew/python/numba ✔                 homebrew/python/pymummer                homebrew/python/scipy ✔                 python3 ✔                             
homebrew/python/cpyrit-cuda             homebrew/python/numpy ✔                 homebrew/python/pyrit                   homebrew/python/vpython                 wxpython                              
homebrew/python/cpyrit-opencl           homebrew/python/pillow ✔                homebrew/python/python-dbus             homebrew/versions/gst-python010         zpython                               
Error: GitHub 401 Unauthorized
HOMEBREW_GITHUB_API_TOKEN may be invalid or expired, check:
  https://github.com/settings/tokens
```

`HOMEBREW_GITHUB_API_TOKEN` is unset. I got several prompts about whether to authorize git-credential-osxkeychain the first time I ran it; I elected to authorize each of them.

Originally reported here: https://groups.google.com/forum/#!msg/homebrew-discuss/fZNmz1h70KY/vCYDXDdDIwAJ

Suspect this might be fallout from #49846 and/or #49954.

### PR DESCRIPTION
**Pull request by tdsmith**
Fixes #50050.

### CODE REVIEW

File: Library/Homebrew/utils.rb [hunk 1]
```diff
@@ -498,11 +498,19 @@ def pretty_ratelimit_reset(reset)
 
   class AuthenticationFailedError < Error
     def initialize(error)
-      super <<-EOS.undent
-        GitHub #{error}
-        HOMEBREW_GITHUB_API_TOKEN may be invalid or expired, check:
+      message = "Github #{error}\n"
```
Comment (mikemcquaid) [inline Library/Homebrew/utils.rb]: `GitHub`

Comment (mikemcquaid) [inline Library/Homebrew/utils.rb]: Also, maybe we include the actual message from the exception?

File: Library/Homebrew/utils.rb [hunk 2]
```diff
@@ -498,11 +498,19 @@ def pretty_ratelimit_reset(reset)
...
+      if ENV["HOMEBREW_GITHUB_API_TOKEN"]
+        message << <<-EOS.undent
+          HOMEBREW_GITHUB_API_TOKEN may be invalid or expired; check:
           #{Tty.em}https://github.com/settings/tokens#{Tty.reset}
-      EOS
+        EOS
+      else
+        message << <<-EOS.undent
+          You may have old Github credentials saved in the OS X Keychain
```
Comment (mikemcquaid) [inline Library/Homebrew/utils.rb]: `GitHub`

Comment (mikemcquaid) [inline Library/Homebrew/utils.rb]: Maybe just worth saying ```
Your GitHub credentials in the OS X keychain are invalid.
Clear them with printf "protocol=https\nhost=github.com\n" | git credential-osxkeychain erase
```

Alternatively, I like the idea of perhaps just retrying the API call without the token? Thoughts?

Comment (mikemcquaid) [inline Library/Homebrew/utils.rb]: Actually thinking about it more: ideal case would be to just clear this automatically and retry if this fails on an API call to a public repo?

File: Library/Homebrew/utils.rb (hunk 1 above)
Comment (tdsmith) [inline Library/Homebrew/utils.rb]: In this case I think it's not very enlightening:

```
irb(main):023:0> headers = {"Accept" => "application/vnd.github.v3+json", "Authorization" => "token 12345"}
=> {"Accept"=>"application/vnd.github.v3+json", "Authorization"=>"token 12345"}
irb(main):024:0> begin
irb(main):025:1*   f = Kernel.open("https://api.github.com/users", headers)
irb(main):026:1> rescue OpenURI::HTTPError => e
irb(main):027:1>   puts e.message
irb(main):028:1>   puts e.io.read
irb(main):029:1> end
401 Unauthorized
{"message":"Bad credentials","documentation_url":"https://developer.github.com/v3"}
```

`error` contains "401 Unauthorized"; the JSON is the same whether there's a bogus token or a bogus username/password so I think it isn't very compelling to do something else here.

**Comment by tdsmith** [comment]
Took a crack at this; testing is annoying because I run out of rate limit :p

File: Library/Homebrew/utils.rb [hunk 3]
```diff
@@ -487,7 +487,7 @@ def initialize(reset, error)
         GitHub #{error}
         Try again in #{pretty_ratelimit_reset(reset)}, or create a personal access token:
           #{Tty.em}https://github.com/settings/tokens/new?scopes=&description=Homebrew#{Tty.reset}
-        and then set the token as: HOMEBREW_GITHUB_API_TOKEN
+        and then set the token as: export HOMEBREW_GITHUB_API_TOKEN="your_new_token"
```
Comment (mikemcquaid) [inline Library/Homebrew/utils.rb]: I can't tell by looking but make sure this isn't more than 80 chars wide

**Comment by mikemcquaid** [comment]
I think it could/should autoclear bad credentials in that case. Don't worry about it for now, though; feel free to just strip out the `try_auth` stuff back to what you had before with printing instructions and you can merge it (I'll try to grab the credentials stuff later).

**Comment by tdsmith** [comment]
Will do. FWIW, modifying the keychain feels bad to me but I can't articulate any probable harms.

**Comment by mikemcquaid** [comment]
@tdsmith Yeh, I get that but in the case that the API token can't be used to query a public resource on a public repo: we know those credentials are bad.

File: Library/Homebrew/utils.rb (hunk 3 above)
Comment (tdsmith) [inline Library/Homebrew/utils.rb]: 76!

**Comment by tdsmith** [comment]
Squashed and merged as 41998df.

**Pull request closed without merging**